[WIP] Paypal Integration

pvighi opened the pull request:
This PR adds a new Payment Methods AB test that determines if the user sees a PayPal payment button in addition to the regular card payment flow.
Client side code will probably we redone to work with the new react version of the site so not much effort was put into it (but should still work).

This branch has the PayPal AB test disabled so the button should never be visible but there are some differences:
1. The initial "next" button now says "contribute with card" and is grey instead of green
2. On mobile the payment details form is hidden initially and the "contribute with card" is shown instead, which must be clicked to reveal the form. This behaviour is not desired unless the PayPal option is enabled

pvighi committed:
paypal WIP

pvighi committed:
Merge branch 'master' into paypal

pvighi committed:
copy amount from main form to paypal before making payment and move button to a better location

pvighi committed:
Merge branch 'master' into paypal

pvighi committed:
WIP

pvighi committed:
paypal webhook WIP

pvighi committed:
updated with latest from master branch

pvighi committed:
paypal refactor

pvighi committed:
paypal: refactor, locale specific thank you page redirect and env specific configuration

pvighi committed:
Paypal: small refactor and check to avoid contributions over the max amount

pvighi committed:
paypal: minor refactors, some styling fixes and svg logo

pvighi committed:
paypal: added error page WIP

pvighi committed:
removed debug output in javascript console

pvighi committed:
merge with latests changes from master

pvighi committed:
Paypal: generate a transaction id to store as custom data with payments

pvighi committed:
paypal: remove formatters class

pvighi committed:
Paypal: typed error codes and some refactoring

pvighi committed:
paypal: small refactor

pvighi committed:
paypal: small refactor and updating error message

pvighi committed:
added paypal auth ajax endpoint (not being used yet)

pvighi committed:
disable paypal attest just in case

pvighi committed:
payment services small refactor

alexduf commented on app/controllers/PaymentError.scala:
```diff
@@ -0,0 +1,15 @@
+}
+
+case object PaypalError extends PaymentError {
+  override def message = "Your contribution using PayPal could not be processed. Please try again or use pay by card."
```
Your contribution using PayPal could not be processed. Please try again or pay by card.

pvighi replied:
I thought "use pay by card" was strange as well, but that's the text I've been given. 
I would have written "use another payment method" but maybe that's not specific enough. Even if we keep this line it might me more correct to say "use 'contribute with card'" to match the label in the button

alexduf replied:
ah, fair enough then!

alexduf commented on app/controllers/PaypalController.scala:
```diff
@@ -0,0 +1,105 @@
+    paypalForm.bindFromRequest().fold[Result](
+      hasErrors = form => handleError(CountryGroup.UK, form.errors.mkString(",")),
+      success = form => {
+        val transactionId = transactionUtil.newTransactionId
```
Do you mind if we call it contributionId? that's how I called it in the database.

alexduf commented on app/controllers/PaypalController.scala:
```diff
@@ -0,0 +1,105 @@
+      success = form => {
+        val transactionId = transactionUtil.newTransactionId
+        val paypalService = paymentServices.paypalServiceFor(request)
+        val maxAllowedAmount = configuration.Payment.maxAmountFor(form.countryGroup.currency)
```
`import configuration.Payment`

alexduf commented on app/utils/TransactionUtil.scala:
```diff
@@ -0,0 +1,10 @@
+import java.util.UUID
+
+trait TransactionUtil {
+  def newTransactionId: String
```
Should this attribute keep the type UUID?

pvighi committed:
hack to make payment details visible on mobile when only card variant is selected

pvighi committed:
check that payment state is not failed after executing it

pvighi committed:
renamed transaction id to contribution id to be consistent with Alex's work

pvighi committed:
removed useless code

Commit:
models

alexduf commented on app/controllers/PaypalController.scala:
```diff
@@ -0,0 +1,103 @@
+import scala.util.Right
+
+
+class PaypalController(ws: WSClient, paymentServices: PaymentServices, contributionIdGenerator :ContributionIdGenerator) extends Controller {
```
I'm being pedantic but the spacing for the parameter contributionIdGenerator is wrong

pvighi committed:
minor formatting fix

Commit:
add a controller to receive the hook

pvighi committed:
verify that the status is approved when executing paypal payments

Commit:
configure the database

Commit:
add insertion of the hook in the database

pvighi committed:
changed style for the control variant of payment methods..

to look exactly like the current version of the frontend in prod. This
way we can disable the paypal variant and we can merge this without
introducing any visible changes

Commit:
typo

alexduf committed:
Merge pull request #76 from guardian/paypal-hook

Paypal hook

Commit:
fix sl4j by excluding paypal's transitive dependency

pvighi committed:
Merge branch 'master' into paypal

pvighi committed:
added PayPal web hook id to dev config

Commit:
differentiate json and stringified json

Commit:
add the metadata model and its related upsert

Commit:
add invisible fields for the paypal payment

Commit:
save metadata when the payment is executed

Commit:
add contributor model and associated upsert

Commit:
add ability to decode Identity's cookie

Commit:
save contributor upon payment execution

Commit:
refactor the method structure

Commit:
fixup identity cookie

Commit:
remove wrong table schema

Commit:
remove useless paypal call

alexduf committed:
Merge pull request #78 from guardian/paypal-meta-data

Paypal meta data
